lib/libc/minimal/source/math/sqrt.c

```diff
@@ -5,19 +5,50 @@
  */
 
 #include <math.h>
+#include <stdint.h>
+#include <zephyr/sys/util.h>
 
-double sqrt(double value)
+#define MAX_D_ITTERATIONS	8  /* usually converges in 5 loops */
+				   /* this ensures we break out of the loop */
+#define MAX_D_ERROR_COUNT	5  /* when result almost converges, stop */
+#define EXP_MASK64	GENMASK64(62, 52)
+
+double sqrt(double square)
 {
-	double sqrt = value / 3;
 	int i;
+	int64_t	exponent;
+	double root;
+	double last;
+	int64_t *p_square = (int64_t *)&square;
+	int64_t *p_root = (int64_t *)&root;
+	int64_t *p_last = (int64_t *)&last;
 
-	if (value <= 0) {
-		return 0;
+	if (square == 0.0) {
+		return square;
+	}
+	if (square < 0.0) {
+		return (square - square) / (square - square);
 	}
 
-	for (i = 0; i < 6; i++) {
-		sqrt = (sqrt + value / sqrt) / 2;
+	/* we need a good starting guess so that this will converge quickly,
+	 * we can do this by dividing the exponent part of the float by 2
+	 * this assumes IEEE-754 format doubles
+	 */
+	exponent = ((*p_square & EXP_MASK64)>>52)-1023;
+	if (exponent == 0x7FF-1023) {
+		/* the number is a NAN or inf, return NaN or inf */
+		return square + square;
 	}
+	exponent /= 2;
+	*p_root = (*p_square & ~EXP_MASK64) | (exponent+1023)<<52;
 
-	return sqrt;
+	for (i = 0; i < MAX_D_ITTERATIONS; i++) {
+		last = root;
+		root = (root + square / root) * 0.5;
+		/* if (llabs(*p_root-*p_last)<MAX_D_ERROR_COUNT) break; */
+		if ((*p_root ^ *p_last) < MAX_D_ERROR_COUNT) {
```

Comment:
How did you select the termination value here? Given that the function converges quadratically, if you could get even one or two bits right at the start, along with the exponent, I think you'd converge every time in 6 iterations and could get rid of this extra termination condition entirely.

Reply:
In general due to the early  termination condition, the loop usually exits in 4 loops, sometimes 5. In all of my test cases it never makes it to 6 loops. Exiting from the loop due to reaching the maximum count should never happen, but I felt that there might be a case where the algorithm never converges and toggles between two values. This would be bad since it would hang the machine in an infinite loop.

I was really concerned when using a simple equality test that there would be case where the LSB toggled back and forth because the root was not exact. I really want ensure that we get out of the loop as early as possible.

The EXOR function requires all of the bits above the last little bits be identical. The commented out line above does the same thing but takes a lot more machine cycles and code. We already took care of he possibility of opposite signs above when we ensured that square was greater than zero.

Reply:
@keith-packard did you look at the test cases? I have an issue on two of the twister tests, it appears that some machines fail when there is a floating point operation because they don't support floating point (or maybe NaN and Inf break their FPUs). 
I need to add something to the test cases so they get skipped on architectures with no floating point. Something like:

```
#define IS_MINIMAL_LIBC_NOFP (IS_ENABLED(CONFIG_MINIMAL_LIBC) \
              && !IS_ENABLED(CONFIG_CBPRINTF_FP_SUPPORT))

       if (IS_MINIMAL_LIBC_NOFP
            || !(IS_ENABLED(CONFIG_FPU)
                 || IS_ENABLED(CONFIG_BOARD_NATIVE_POSIX))) {
                ztest_test_skip();
                return;
        }
```
I took this example from tests/lib/sprintf/src/main.c but I wasn.t sure what the conditions for skipping the test should be.

Reply:
You can check for a IEC-60559 conforming C environment by `#include <float.h>` and then using `#ifdef __STDC_IEC_559__`. I think that was added in C99, so it should be set correctly by any compiler supported by Zephyr.